### PR TITLE
WIP #776 WS symbol data implementation

### PR DESCRIPTION
**Pull request by Werdor**
- Realtime candle data
- recent trades
- order book

depends on #780

### CODE REVIEW

File: trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            for time_frame in self.time_frames:
+                self._add_feed_and_run_if_required(
+                    TRADES,
+                    OHLCV(Callback(OHLCVCallBack(self, time_frame).ohlcv_callback),
+                          window=self._convert_time_frame_minutes_to_seconds(TimeFramesMinutes[time_frame])))
+
+            # add real time handler
+            self._add_feed_and_run_if_required(
+                TRADES,
+                OHLCV(Callback(OHLCVCallBack(self, 0).ohlcv_callback), window=0))
```
Comment (GuillaumeDSM) [inline trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py]: is it necessary ? i don't understand why we need a 0 timeframe

File: trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+                self.is_handling_price_ticker or \
+                self.is_handling_funding or \
+                self.is_handling_ohlcv or \
+                self.is_handling_recent_trades:
+            self.cryptofeed_handler.run()
+            is_websocket_running = True
+
+        if not is_websocket_running:
+            self.logger.error(f"{self.exchange_manager.exchange_class_string.title()}'s "
+                              f"websocket is handling anything, it will not be started, ")
```
Comment (GuillaumeDSM) [inline trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py]: websocket is *not* handling anything

File: trading/exchanges/websockets_exchanges/websocket/cryptofeed_callbacks.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    def __init__(self, parent):
+        super().__init__(parent)
+
+    async def tickers_callback(self, feed, pair, bid, ask, timestamp):
+        # self.logger.info(f'TICKER / Pair: {pair} Bid: {bid} Ask: {ask}')
+        # symbol_data = self.parent.get_symbol_data_from_pair(pair)
+        # if symbol_data:
+        #     symbol_data.update_symbol_ticker(
+        #         self.parent.convert_into_ccxt_ticker(symbol_data.symbol, bid, ask, timestamp))
+        pass
```
Comment (GuillaumeDSM) [inline trading/exchanges/websockets_exchanges/websocket/cryptofeed_callbacks.py]: to be implemented later ?

File: trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py (hunk 1 above)
Comment (Werdor) [reply trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py]: It's not a 0 timeframe, cryptofeed provide "windows" that corresponds to seconds between updates. So here I specified 0s to handle all updates.

File: trading/exchanges/websockets_exchanges/websocket/cryptofeed_callbacks.py (hunk 3 above)
Comment (Werdor) [reply trading/exchanges/websockets_exchanges/websocket/cryptofeed_callbacks.py]: I am working on it, it's the reason of the "WIP" ;)

File: trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        self.is_handling_order_book = False
+        self.is_handling_recent_trades = False
+        self.is_handling_funding = False
+
+        self.channels = []
+        self.callbacks = {}
+
+    def init_web_sockets(self, time_frames, trader_pairs):
+        self.cryptofeed_handler = FeedHandler()
+        self.exchange_class = self.__class__._get_cryptofeed_class(self.exchange_manager.exchange_class_string.title())
```
[the remark this answers is not in the document]
Comment (Werdor) [reply trading/exchanges/websockets_exchanges/websocket/cryptofeed_websocket.py]: :+1:

**Comment by Herklos** [comment]
https://github.com/Drakkar-Software/OctoBot/pull/797

**Pull request closed without merging**